experimental/transliteration/src/transliterator/mod.rs

```diff
@@ -0,0 +1,1417 @@
+                    dest.curr_replacement_len(),
+                    0,
+                    "pre-start cursor not the first replacement"
+                );
+                dest.set_offset_to_chars_off_start(num);
+            }
+            &Self::BackReference(num) => {
+                dest.push_str(data.get_segment(num as usize));
+            }
+            Self::FunctionCall(call) => {
+                // the way function call replacing works is as such:
+                // use the same backing buffer as the parent Replaceable, but set
+                // the visible content range of the Replaceable appropriately.
+
+                // eprintln!("dest before function call: {dest:?}");
+
+                let mut range_aggregator = dest.start_replaceable_adapter();
+                helpers::replace_encoded_str(&call.arg, &mut range_aggregator, data, vt, env);
+
+                call.translit
+                    .transliterate(range_aggregator.as_replaceable().child(), env);
+            }
+        }
+    }
+}
+
+enum VarTableElement<'a> {
+    Compound(&'a str),
+    Quantifier(QuantifierKind, &'a str),
+    Segment(Segment<'a>),
+    UnicodeSet(CodePointInversionListAndStringList<'a>),
+    FunctionCall(FunctionCall<'a>),
+    BackReference(u16),
+    AnchorStart,
+    AnchorEnd,
+    LeftPlaceholderCursor(u16),
+    RightPlaceholderCursor(u16),
+    PureCursor,
+}
+
+impl<'a> VarTableElement<'a> {
+    fn into_replacer(self) -> Option<SpecialReplacer<'a>> {
+        Some(match self {
+            Self::Compound(elt) => SpecialReplacer::Compound(elt),
+            Self::FunctionCall(elt) => SpecialReplacer::FunctionCall(elt),
+            Self::BackReference(elt) => SpecialReplacer::BackReference(elt),
+            Self::LeftPlaceholderCursor(elt) => SpecialReplacer::LeftPlaceholderCursor(elt),
+            Self::RightPlaceholderCursor(elt) => SpecialReplacer::RightPlaceholderCursor(elt),
+            Self::PureCursor => SpecialReplacer::PureCursor,
+            _ => return None,
+        })
+    }
+
+    fn into_matcher(self) -> Option<SpecialMatcher<'a>> {
+        Some(match self {
+            Self::Compound(elt) => SpecialMatcher::Compound(elt),
+            Self::Quantifier(kind, elt) => SpecialMatcher::Quantifier(kind, elt),
+            Self::Segment(elt) => SpecialMatcher::Segment(elt),
+            Self::UnicodeSet(elt) => SpecialMatcher::UnicodeSet(elt),
+            Self::AnchorEnd => SpecialMatcher::AnchorEnd,
+            Self::AnchorStart => SpecialMatcher::AnchorStart,
+            _ => return None,
+        })
+    }
+}
+
+impl<'a> VarTable<'a> {
+    // TODO(#3958): these must be the same as during datagen. Find some place to define them *once*
+    const BASE: u32 = '\u{F0000}' as u32;
+    const MAX_DYNAMIC: u32 = '\u{FFFF0}' as u32;
+    const RESERVED_PURE_CURSOR: u32 = '\u{FFFFB}' as u32;
+    const RESERVED_ANCHOR_START: u32 = '\u{FFFFC}' as u32;
+    const RESERVED_ANCHOR_END: u32 = '\u{FFFFD}' as u32;
```

Comment:
I'm growing reserved chars from top-to-bottom because the dynamically sized stuff is growing from bottom-to-top, and I'm leaving some space to `MAX_DYNAMIC`